packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol

```diff
@@ -595,3 +595,85 @@ contract L1CrossDomainMessenger_Test is Bridge_Initializer {
         assertEq(l1CrossDomainMessenger.paused(), superchainConfig.paused());
     }
 }
+
+/// @dev A test demonstarting a reentrancy vulnerability in the CrossDomainMessenger contract, which
+///      is possible by intercepting and sandwhiching a signed Safe Transaction to upgrade it.
+contract L1CrossDomainMessenger_ReinitReentryTest is Bridge_Initializer {
+    bool attacked;
+
+    // Common values used across functions
+    uint256 constant messageValue = 50;
+    bytes constant selector = abi.encodeWithSelector(this.reinitAndReenter.selector);
+    address sender;
+    bytes32 hash;
+    address target;
+
+    function setUp() public override {
+        super.setUp();
+        target = address(this);
+        sender = Predeploys.L2_CROSS_DOMAIN_MESSENGER;
+        hash = Hashing.hashCrossDomainMessage(
+            Encoding.encodeVersionedNonce({ _nonce: 0, _version: 1 }), sender, target, messageValue, 0, selector
+        );
+        vm.deal(address(l1CrossDomainMessenger), messageValue * 2);
+    }
+
+    /// @dev This method will be called by the relayed message, and will attempt to reenter the relayMessage function
+    ///      exactly once.
+    function reinitAndReenter() public payable {
+        // only attempt the attack once
+        if (!attacked) {
+            attacked = true;
```

Comment:
Addresses [this comment](https://github.com/ethereum-optimism/optimism/pull/8797#discussion_r1442358652):

> https://github.com/ethereum-optimism/optimism/pull/8797#discussion_r1442358652